### ISSUE TITLE
Unsupported mode 120000 panic while running 'git ls-tree'

### ISSUE THREAD

**Issue by sirnewton01** [driusan/dgit#15]
I checked out git from GitHub using the standard git tool: https://github.com/git/git

When I run 'git ls-tree' on a specific commit in that repository I get an error:

```
$ git ls-tree e3331758f12da22f4103eec7efe1b5304a9be5e9

panic: Unsupported mode 120000 in tree 313f70847d0dab2718d19201b5be3af52061c4da

goroutine 1 [running]:
github.com/driusan/dgit/git.TreeID.GetAllObjects(0x27ab0d7d84703f31, 0xf53abeb50192d118, 0xdac46120, 0xc42009ad80, 0x0, 0x0, 0x0, 0x0, 0x27ab0d7d84703f31, 0xf53abeb50192d118)
	/Users/cmcgee/go/src/github.com/driusan/dgit/git/sha1.go:482 +0x92f
github.com/driusan/dgit/git.expandGitTreeIntoIndexesRecursive(0xc42009ad80, 0x27ab0d7d84703f31, 0xf53abeb50192d118, 0xc4dac46120, 0x0, 0x0, 0xdac40000, 0x0, 0x0, 0x1, ...)
	/Users/cmcgee/go/src/github.com/driusan/dgit/git/expandtree.go:38 +0x76
github.com/driusan/dgit/git.expandGitTreeIntoIndexes(0xc42009ad80, 0x13746e0, 0xc423be79e0, 0x100000001010000, 0x15, 0x30, 0xc42009e1b0, 0xc423fb7460, 0x10c8795)
	/Users/cmcgee/go/src/github.com/driusan/dgit/git/expandtree.go:26 +0x15a
github.com/driusan/dgit/git.lsTree(0xc42009ad80, 0x1, 0x28, 0x0, 0x13746e0, 0xc423be79e0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/cmcgee/go/src/github.com/driusan/dgit/git/lstree.go:161 +0x72
github.com/driusan/dgit/git.LsTree(0xc42009ad80, 0x1, 0x28, 0x0, 0x13746e0, 0xc423be79e0, 0x0, 0x0, 0x0, 0x11dfd25, ...)
	/Users/cmcgee/go/src/github.com/driusan/dgit/git/lstree.go:61 +0x11e5
github.com/driusan/dgit/cmd.LsTree(0xc42009ad80, 0xc42009a0b0, 0x1, 0x1, 0xc42009ad80, 0x0)
	/Users/cmcgee/go/src/github.com/driusan/dgit/cmd/lstree.go:47 +0x699
main.main()
	/Users/cmcgee/go/src/github.com/driusan/dgit/main.go:178 +0x2acd
```

**Comment by driusan** [comment]
Probably related to #14, that's the mode git uses for symbolic links.

**Comment by sirnewton01** [comment]
If you want I can close this issue. #14 is a kind of catch-all for "symbolic links aren't working"

**Comment by driusan** [comment]
No, leave them both, they're probably different changes to fix them, but they're just both related to the underlying issues of symlinks not working.

**Comment by driusan** [comment]
I've fixed the crash in ls-tree for symlinks in 8cb0a3f, but tree 313f70847d0dab2718d19201b5be3af52061c4da also has this entry:

160000 commit 19d97bf5af05312267c2e874ee6bcf584d9e9681  sha1collisiondetection

which is causing problems, and making dgit error out with an "Object not found" error. (It's also really slow. In fact, there's no reason to be looking up the object at all for ls-tree, it's only because there's some plumbing shared with other commands that need to know the size of the tree object.)

**Comment by driusan** [comment]
Crashes are fixed and the performance is reasonable now (but still worse than the official git client).. if you do a diff against the two it'll say there's a few differences because (at least on my machine) git is using an insane collation that goes:

1. foo.x
2. foo
3. fooa

Rather than:

1. foo
2. foo.x
3. fooa

while dgit is just using Go's `<` operator to sort the paths, but `dgit ls-tree` now works on weirdo trees like that one and doesn't die on symlinks.
